File: frontend/src/shared/api/bandit.ts
import { apiClient } from "@/shared/api/serverClient";

export interface BanditScanTask {
  id: string;
  project_id: string;
  name: string;
  status: string;
  target_path: string;
  severity_level: string;
  confidence_level: string;
  total_findings: number;
  high_count: number;
  medium_count: number;
  low_count: number;
  scan_duration_ms: number;
  files_scanned: number;
  error_message?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export interface BanditFinding {
  id: string;
  scan_task_id: string;
  test_id: string;
  test_name?: string | null;
  issue_text?: string | null;
  issue_severity: string;
  issue_confidence: string;
  file_path: string;
  line_number?: number | null;
  code_snippet?: string | null;
  more_info?: string | null;
  status: string;
  created_at?: string;
}

export interface BanditRule {
  id: string;
  test_id: string;
  name: string;
  description: string;
  description_summary: string;
  checks: string[];
  source: string;
  bandit_version: string;
  is_active: boolean;
  created_at?: string | null;
  updated_at?: string | null;
}

export async function createBanditScanTask(params: {
  project_id: string;
  name?: string;
  target_path?: string;
  severity_level?: string;
  confidence_level?: string;
}): Promise<BanditScanTask> {
  const response = await apiClient.post("/static-tasks/bandit/scan", params);
  return response.data;
}

export async function getBanditScanTask(taskId: string): Promise<BanditScanTask> {
  const response = await apiClient.get(`/static-tasks/bandit/tasks/${taskId}`);
  return response.data;
}

export async function interruptBanditScanTask(
  taskId: string,
): Promise<{ message: string; task_id: string; status: string }> {
  const response = await apiClient.post(`/static-tasks/bandit/tasks/${taskId}/interrupt`);
  return response.data;
}

export async function getBanditScanTasks(params?: {
  projectId?: string;
  status?: string;
  skip?: number;
  limit?: number;
}): Promise<BanditScanTask[]> {
  const searchParams = new URLSearchParams();
  if (params?.projectId) searchParams.set("project_id", params.projectId);
  if (params?.status) searchParams.set("status", params.status);
  if (params?.skip !== undefined) searchParams.set("skip", String(params.skip));
  if (params?.limit !== undefined) searchParams.set("limit", String(params.limit));
  const query = searchParams.toString();
  const response = await apiClient.get(
    `/static-tasks/bandit/tasks${query ? `?${query}` : ""}`,
  );
  return response.data;
}

export async function getBanditFindings(params: {
  taskId: string;
  status?: string;
  skip?: number;
  limit?: number;
}): Promise<BanditFinding[]> {
  const searchParams = new URLSearchParams();
  if (params.status) searchParams.set("status", params.status);
  if (params.skip !== undefined) searchParams.set("skip", String(params.skip));
  if (params.limit !== undefined) searchParams.set("limit", String(params.limit));
  const query = searchParams.toString();
  const response = await apiClient.get(
    `/static-tasks/bandit/tasks/${params.taskId}/findings${query ? `?${query}` : ""}`,
  );
  return response.data;
}

export async function getBanditFinding(params: {
  taskId: string;
  findingId: string;
}): Promise<BanditFinding> {
  const response = await apiClient.get(
    `/static-tasks/bandit/tasks/${params.taskId}/findings/${params.findingId}`,
  );
  return response.data;
}

export async function updateBanditFindingStatus(params: {
  findingId: string;
  status: "open" | "verified" | "false_positive" | "fixed";
}): Promise<{ message: string; finding_id: string; status: string }> {
  const response = await apiClient.post(
    `/static-tasks/bandit/findings/${params.findingId}/status`,
    undefined,
    { params: { status: params.status } },
  );
  return response.data;
}

export async function getBanditRules(params?: {
  is_active?: boolean;
  source?: string;
  keyword?: string;
  skip?: number;
  limit?: number;
}): Promise<BanditRule[]> {
  const searchParams = new URLSearchParams();
  if (params?.is_active !== undefined) {
    searchParams.set("is_active", String(params.is_active));
  }
  if (params?.source) searchParams.set("source", params.source);
  if (params?.keyword) searchParams.set("keyword", params.keyword);
  if (params?.skip !== undefined) searchParams.set("skip", String(params.skip));
  searchParams.set("limit", String(params?.limit ?? 1000));
  const query = searchParams.toString();
  const response = await apiClient.get(
    `/static-tasks/bandit/rules${query ? `?${query}` : ""}`,
  );
  return response.data;
}

export async function getBanditRule(ruleId: string): Promise<BanditRule> {
  const response = await apiClient.get(`/static-tasks/bandit/rules/${ruleId}`);
  return response.data;
}

export async function updateBanditRuleEnabled(params: {
  ruleId: string;
  is_active: boolean;
}): Promise<{ message: string; rule_id: string; is_active: boolean }> {
  const response = await apiClient.post(
    `/static-tasks/bandit/rules/${params.ruleId}/enabled`,
    { is_active: params.is_active },
  );
  return response.data;
}

export async function batchUpdateBanditRulesEnabled(params: {
  rule_ids?: string[];
  source?: string;
  keyword?: string;
  current_is_active?: boolean;
  is_active: boolean;
}): Promise<{ message: string; updated_count: number; is_active: boolean }> {
  const response = await apiClient.post(`/static-tasks/bandit/rules/batch/enabled`, params);
  return response.data;
}

File: backend_old/app/models/bandit.py
"""
Bandit 静态扫描模型

用途：
- 持久化 Bandit 静态扫描任务元数据（bandit_scan_tasks）
- 持久化 Bandit 扫描发现明细（bandit_findings）
- 持久化 Bandit 规则启停状态（bandit_rule_states）
"""

import uuid
from sqlalchemy import Column, String, Integer, DateTime, ForeignKey, Text, Index, Boolean
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship

from app.models.base import Base


class BanditScanTask(Base):
    """Bandit 扫描任务"""

    __tablename__ = "bandit_scan_tasks"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = Column(String, ForeignKey("projects.id"), nullable=False)
    name = Column(String, nullable=False)
    status = Column(
        String,
        default="pending",
        comment="pending, running, completed, failed, interrupted",
    )
    target_path = Column(String, nullable=False, comment="扫描目标路径")
    severity_level = Column(String, default="medium", comment="low, medium, high")
    confidence_level = Column(String, default="medium", comment="low, medium, high")
    total_findings = Column(Integer, default=0, comment="发现总数")
    high_count = Column(Integer, default=0, comment="HIGH 严重度数量")
    medium_count = Column(Integer, default=0, comment="MEDIUM 严重度数量")
    low_count = Column(Integer, default=0, comment="LOW 严重度数量")
    scan_duration_ms = Column(Integer, default=0, comment="扫描耗时(毫秒)")
    files_scanned = Column(Integer, default=0, comment="扫描文件数")
    error_message = Column(Text, nullable=True, comment="失败/中断错误信息")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    __table_args__ = (
        Index("ix_bandit_tasks_project_created_at", "project_id", created_at.desc()),
        Index(
            "ix_bandit_tasks_project_lower_status_created_at",
            "project_id",
            func.lower(status),
            created_at.desc(),
        ),
    )

    project = relationship("Project", back_populates="bandit_scan_tasks")
    findings = relationship(
        "BanditFinding", back_populates="scan_task", cascade="all, delete-orphan"
    )


class BanditFinding(Base):
    """Bandit 扫描发现"""

    __tablename__ = "bandit_findings"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    scan_task_id = Column(String, ForeignKey("bandit_scan_tasks.id"), nullable=False)
    test_id = Column(String, nullable=False, comment="Bandit 规则ID，例如 B602")
    test_name = Column(String, nullable=False, comment="Bandit 规则名称")
    issue_severity = Column(String, nullable=False, comment="HIGH, MEDIUM, LOW")
    issue_confidence = Column(String, nullable=False, comment="HIGH, MEDIUM, LOW")
    file_path = Column(String, nullable=False)
    line_number = Column(Integer, nullable=True)
    code_snippet = Column(Text, nullable=True)
    issue_text = Column(Text, nullable=True)
    more_info = Column(String, nullable=True)
    status = Column(String, default="open", comment="open, verified, false_positive")
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    __table_args__ = (
        Index(
            "ix_bandit_findings_scan_task_status_created",
            "scan_task_id",
            "status",
            created_at.desc(),
        ),
        Index(
            "ix_bandit_findings_scan_task_file_line",
            "scan_task_id",
            "file_path",
            "line_number",
        ),
        Index(
            "ix_bandit_findings_scan_task_severity",
            "scan_task_id",
            "issue_severity",
        ),
    )

    scan_task = relationship("BanditScanTask", back_populates="findings")


class BanditRuleState(Base):
    """Bandit 规则启停状态（仅用于前端规则页展示，不影响扫描命令）。"""

    __tablename__ = "bandit_rule_states"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    test_id = Column(String, nullable=False, unique=True, comment="Bandit 规则ID，例如 B602")
    is_active = Column(Boolean, nullable=False, default=True, comment="规则是否启用")
    is_deleted = Column(Boolean, nullable=False, default=False, comment="规则是否软删除")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    __table_args__ = (
        Index("ix_bandit_rule_states_test_id", "test_id"),
        Index("ix_bandit_rule_states_is_active", "is_active"),
        Index("ix_bandit_rule_states_is_deleted", "is_deleted"),
    )
